week_3/Binary_Tree/class.js

```diff
@@ -0,0 +1,117 @@
+class Node {
+	constructor(val) {
+		this.val = val;
+		this.left = null;
+		this.right = null;
+	}
+}
+class BinarySearchTree {
+	constructor() {
+		this.root = null;
+	}
+	BFS() {
+		if (!this.root) {
+			return null;
+		}
+		let nodes = [];
+		let queue = [];
+		queue.push(this.root);
+
+		while (queue.length) {
+			let current = queue.shift();
+			nodes.push(current.val);
+			if (current.left) queue.push(current.left);
+			if (current.right) queue.push(current.right);
+		}
+		return nodes;
+	}
+	DFSPostOrder() {
+		if (!this.root) return null;
+		let visited = [];
+
+		const visit = (node) => {
+			if (node.left) visit(node.left);
+			if (node.right) visit(node.right);
+			visited.push(node.val);
+		};
+		visit(this.root);
+		return visited;
+	}
+	DFSInOrder() {
+		if (!this.root) return null;
+		let visited = [];
+		const traverse = (node) => {
+			if (node.left) traverse(node.left);
+			visited.push(node.val);
+			if (node.right) traverse(node.right);
+		};
+		traverse(this.root);
+		return visited;
+	}
+	DFSPreOrder() {
+		if (!this.root) return null;
+		let visited = [];
+
+		const visit = (node) => {
+			visited.push(node.val);
+			if (node.left) visit(node.left);
+			if (node.right) visit(node.right);
+		};
+		visit(this.root);
+		return visited;
+	}
+	insert(val) {
+		let newNode = new Node(val);
+		if (!this.root) {
+			this.root = newNode;
+			return this;
+		}
+		let comp = this.root;
+		while (true) {
+			if (val === comp.val) return undefined;
+			if (comp.val < val) {
+				if (!comp.right) {
+					comp.right = newNode;
+					return this;
+				}
+				comp = comp.right;
+			} else {
+				if (!comp.left) {
+					comp.left = newNode;
+					return this;
+				}
+				comp = comp.left;
+			}
+		}
+	}
+	find(val) {
+		if (!this.root) return false;
+		let current = this.root,
```

Comment:
Is this comma intentional?

Reply:
Yes. This is just a way to assign multiple related variables at once without having to use "let" over and over -- which admittedly isn't especially useful when there are only two variables.

Reply:
Yeah, I'd argue it makes it more difficult to read actually; can we swap this out with new variables on each line? This is the type of thing you'd find in "code golf" solutions where one tries to accomplish solving a problem with as little code as possible, but it drastically reduces readability in real code bases that are meant to stand the test of time